[4.6] media-libs/mlt-0.8.8 ICE with -fprefetch-loop-arrays in build2_stat, at tree.c

https://bugs.gentoo.org/454568
http://gcc.gnu.org/PR53550

commit e16a56e1c204eda1644c6003fee840a7a6abcc3b
Author: jakub <jakub@138bc75d-0d04-0410-961f-82ee72b054a4>
Date:   Mon Jun 4 09:27:00 2012 +0000

    	PR tree-optimization/53550
    	* tree-ssa-loop-niter.c (number_of_iterations_cond): If type
    	is POINTER_TYPE_P, use sizetype as step type instead of type.
    
    	* gcc.dg/pr53550.c: New test.
    
    
    git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_7-branch@188170 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/ChangeLog                  |  6 ++++++
 gcc/testsuite/ChangeLog        |  5 +++++
 gcc/testsuite/gcc.dg/pr53550.c | 12 ++++++++++++
 gcc/tree-ssa-loop-niter.c      |  7 ++++---
 4 files changed, 27 insertions(+), 3 deletions(-)

--- /dev/null
+++ b/gcc/testsuite/gcc.dg/pr53550.c
@@ -0,0 +1,12 @@
+/* PR tree-optimization/53550 */
+/* { dg-do compile } */
+/* { dg-options "-O2 -fprefetch-loop-arrays -w" } */
+
+int *
+foo (int *x)
+{
+  int *a = x + 10, *b = x, *c = a;
+  while (b != c)
+    *--c = *b++;
+  return x;
+}
--- a/gcc/tree-ssa-loop-niter.c
+++ b/gcc/tree-ssa-loop-niter.c
@@ -1276,13 +1276,14 @@ number_of_iterations_cond (struct loop *loop,
      practice, but it is simple enough to manage.  */
   if (!integer_zerop (iv0->step) && !integer_zerop (iv1->step))
     {
+      tree step_type = POINTER_TYPE_P (type) ? sizetype : type;
       if (code != NE_EXPR)
 	return false;
 
-      iv0->step = fold_binary_to_constant (MINUS_EXPR, type,
+      iv0->step = fold_binary_to_constant (MINUS_EXPR, step_type,
 					   iv0->step, iv1->step);
       iv0->no_overflow = false;
-      iv1->step = build_int_cst (type, 0);
+      iv1->step = build_int_cst (step_type, 0);
       iv1->no_overflow = true;
     }
 
